File: server/features_test.go
package main

import (
	"encoding/base64"
	"math"
	"testing"
)

func TestAllPowers(t *testing.T) {
	for k := 1; k <= 12; k++ {
		s := initial()
		s.Ball.Owner = 16
		s.pickup(7, k)
		switch k {
		case 7:
			if s.Ball.Owner != 7 {
				t.Fatal("magnet")
			}
		case 8:
			if s.Players[s.Ball.Owner].Team != 0 {
				t.Fatal("transport")
			}
		case 11:
			if s.Players[7].Health != 100 {
				t.Fatal("heal")
			}
		case 12:
			if s.Players[16].Stun <= 0 || s.Ball.Owner != -1 {
				t.Fatal("zap")
			}
		default:
			if s.Effect.Kind != k || s.Effect.Time != 6 || s.Effect.Team != 0 {
				t.Fatal("temporary power", k)
			}
		}
	}
}
func TestPowerMovement(t *testing.T) {
	for _, c := range []struct {
		k, owner int
		factor   float64
	}{{1, 16, 0}, {2, 16, -1}, {3, 16, 1}, {4, 7, 1.4}, {5, 16, 1.4}, {6, 16, 1}} {
		s := isolated()
		s.Ball.H = 4
		s.pickup(c.owner, c.k)
		x := s.Players[7].X
		s.step(dt, [2]Input{{X: 1}, {}})
		if math.Abs(s.Players[7].X-x-5*velocityUnit*dt*c.factor) > 1e-8 {
			t.Fatal("movement", c.k, s.Players[7].X)
		}
	}
}
func TestPowerExpiryShieldHeal(t *testing.T) {
	s := initial()
	s.pickup(7, 1)
	s.pickup(16, 10)
	if s.active(1, -1) || !s.active(10, 1) {
		t.Fatal("replacement")
	}
	if s.damage(7, 16) {
		t.Fatal("shield failed")
	}
	s.pickup(7, 12)
	if s.Players[16].Health != 100 {
		t.Fatal("shield failed against zap")
	}
	s.matchClock(6)
	if s.Effect.Kind != 0 {
		t.Fatal("expiry")
	}
	s.damage(7, 16)
	s.pickup(16, 11)
	if s.Players[16].Health != 100 {
		t.Fatal("heal")
	}
	s.Tick = 1
	s.Players[7].Health = 45
	s.resetPitch()
	if s.Players[7].Health != 45 {
		t.Fatal("goal reset healed injury damage")
	}
}
func TestDoorBothHalves(t *testing.T) {
	for _, period := range []int{1, 2} {
		s := isolated()
		s.Period = period
		s.pickup(7, 9)
		x := -20.9
		if period == 2 {
			x = -x
		}
		s.Ball = Ball{X: x, H: 1, VX: math.Copysign(24, x), Owner: -1, LastTouch: -1}
		s.step(dt, [2]Input{})
		s.step(dt, [2]Input{})
		if s.Score[1] != 0 || s.Ball.VX*x >= 0 {
			t.Fatal("door")
		}
		if s.goalBlocked(-x) {
			t.Fatal("wrong door")
		}
	}
}
func TestWarpsAndHighBounce(t *testing.T) {
	for _, x := range []float64{-8, 8} {
		for _, sign := range []float64{-1, 1} {
			s := isolated()
			s.Ball = Ball{X: x, Z: 11.1 * sign, H: 1, VZ: 24 * sign, Owner: -1, LastTouch: 7}
			s.step(dt, [2]Input{})
			s.step(dt, [2]Input{})
			if s.Ball.Z*sign >= -10 || s.Ball.VZ*sign <= 0 || s.Event.Kind != 12 {
				t.Fatal("warp", s.Ball)
			}
		}
	}
	s := isolated()
	s.Ball = Ball{X: 8, Z: 11.1, H: 3, VZ: 24, Owner: -1, LastTouch: -1}
	s.step(dt, [2]Input{})
	s.step(dt, [2]Input{})
	if s.Ball.VZ >= 0 || s.Ball.Z < 0 {
		t.Fatal("high ball warped")
	}
}
func TestElectroChargesAndShieldCatch(t *testing.T) {
	s := isolated()
	s.Multiplier = 2
	s.Ball = Ball{ElectricBudget: 3, X: 304 * terrainUnit, Z: 11.3, H: 1, LastTouch: 7, Owner: -1}
	s.sideFeature()
	if s.Ball.Electric != 3 {
		t.Fatal("charge")
	}
	s.Players[16] = Player{X: 5, Z: 2, Health: 100, Team: 1, FX: -1}
	s.Ball.X = 5
	s.Ball.Z = 2
	s.step(dt, [2]Input{})
	if s.Ball.Owner != -1 || s.Ball.Electric != 2 || s.Players[16].Stun <= 0 {
		t.Fatal("electro hit")
	}
	s.Players[16].Stun = 0
	s.pickup(16, 10)
	s.Ball.X = s.Players[16].X
	s.Ball.Z = s.Players[16].Z
	s.step(dt, [2]Input{})
	if s.Ball.Owner != 16 || s.Ball.Electric != 0 {
		t.Fatal("shield catch")
	}
}
func TestInjuryMedicalAndNoReserves(t *testing.T) {
	s := initial()
	s.Players[16].Health = 1
	s.damage(7, 16)
	if s.Score[0] != 10 || s.Players[16].Injury != 6 {
		t.Fatal("injury")
	}
	clock := s.Time
	for i := 0; i < 361; i++ {
		s.step(dt, [2]Input{})
	}
	if s.Time != clock || s.Reserves[1] != 2 || s.Players[16].Health != 100 {
		t.Fatal("medical replacement")
	}
	s.Reserves[1] = 0
	s.Players[16].Health = 1
	s.Players[16].Stun = 0
	s.damage(7, 16)
	for i := 0; i < 361; i++ {
		s.step(dt, [2]Input{})
	}
	if s.Players[16].Health != 0 || s.Score[0] != 20 || s.Reserves[1] != 0 {
		t.Fatal("no reserves")
	}
}
func TestPickupRespawnAndEquipment(t *testing.T) {
	s := isolated()
	s.Pickups[0] = Pickup{Kind: 13, X: 0, Z: 0, Life: 14}
	s.featureStep(dt)
	s.featureStep(dt)
	if s.Credits[0] != 10 || s.Pickups[0].Wait <= 0 {
		t.Fatal("double pickup")
	}
	s.pickup(7, 17)
	if movementSpeed(&s.Players[7], false, false) != 7*velocityUnit {
		t.Fatal("boots")
	}
	s.pickup(7, 18)
	s.throw(7, false)
	if s.Ball.VX != 8*velocityUnit || s.Ball.SpeedTimer != 125 {
		t.Fatal("throw gear")
	}
	s.Players[16].Stun = 0
	s.damage(16, 7)
	if s.Players[7].Gear != 0 {
		t.Fatal("gear not lost")
	}
}
func TestFeatureSnapshot(t *testing.T) {
	s := initial()
	s.Effect = Effect{10, 1, 5.5}
	s.Credits = [2]int{100, 50}
	s.Reserves = [2]int{2, 1}
	s.Ball.Charged = true
	s.Ball.Electric = 3
	for i := range s.Players {
		s.Players[i].Health = float64(90 - i)
		s.Players[i].Injury = float64(i) / 10
		s.Players[i].Gear = 14 + i%8
		for j := range s.Players[i].Stats {
			s.Players[i].Stats[j] = 100 + i
		}
	}
	for i := 0; i < 20; i++ {
		s.event(3+i%3, 7, -1, float64(i), -2, 1)
	}
	raw := encodeSnapshot(Snapshot{State: s, Room: "ABCDEF", Names: [2]string{"Blue", "Red"}, Started: true})
	if len(raw) > 1200 {
		t.Fatal("snapshot exceeds datagram", len(raw))
	}
	t.Log("WIRE:" + base64.StdEncoding.EncodeToString(raw))
}

func TestTransportRosterSlot(t *testing.T) {
	for team := 0; team < 2; team++ {
		for period := 1; period <= 2; period++ {
			s := initial()
			s.Period = period
			target := team*9 + 8
			s.Players[target].X = 0
			s.Players[team*9+7].X = 20
			s.pickup(team*9+7, 8)
			if s.Ball.Owner != target {
				t.Fatal("transport must target roster slot eight")
			}
			s.Ball.Owner = 1
			s.Players[target].Stun = 1
			s.pickup(team*9+7, 8)
			if s.Ball.Owner != 1 {
				t.Fatal("fallen target must not trigger fallback")
			}
		}
	}
}

func TestSelectedGroundedPickup(t *testing.T) {
	for _, kind := range []int{1, 13, 17} {
		s := initial()
		for i := range s.Players {
			s.Players[i].X = 10
			s.Players[i].Z = 10
		}
		s.Pickups[0] = Pickup{Kind: kind, Life: 14}
		s.Players[6].X = 0
		s.Players[6].Z = 0
		s.featureStep(dt)
		if s.Pickups[0].Wait != 0 {
			t.Fatal("unselected pickup")
		}
		s.Players[7].X = 0
		s.Players[7].Z = 0
		s.Players[7].Action = 2
		s.featureStep(dt)
		if s.Pickups[0].Wait != 0 {
			t.Fatal("jumping pickup")
		}
		s.Players[7].Action = 0
		s.Players[16].X = 0
		s.Players[16].Z = 0
		s.featureStep(dt)
		if s.Pickups[0].Wait <= 0 {
			t.Fatal("selected player missed pickup")
		}
		if kind == 1 && s.Effect.Team != 0 {
			t.Fatal("power priority")
		}
		if kind == 13 && s.Credits != [2]int{10, 0} {
			t.Fatal("credit priority")
		}
		if kind == 17 && s.Players[7].Gear != 17 {
			t.Fatal("equipment priority")
		}
	}
}

File: src/features.test.js
import { spawnPickup } from "./pickup-spawn.js";
import { movementSpeed, velocityUnit } from "./attributes.js";
import { test } from "node:test";
import assert from "node:assert/strict";
import { matchClock, initial, step, resetPitch, throwBall } from "./game.js";
import {
  pickup,
  active,
  damage,
  featureStep,
  sideFeature,
  goalBlocked,
} from "./features.js";
const dt = 1 / 60;
function isolated() {
  const s = initial();
  s.logicalView = [160, 380];
  s.players.forEach((p) => (p.stun = 100));
  Object.assign(s.players[7], { x: 4, z: 0, stun: 0, fx: 1, fz: 0 });
  s.ball.x = 4;
  s.ball.z = 0;
  s.ball.h = 4;
  return s;
}
test("all twelve powers apply their advertised effect", () => {
  for (let k = 1; k <= 12; k++) {
    const s = initial();
    s.ball.owner = 16;
    pickup(s, 7, k);
    if ([1, 2, 3, 4, 5, 6, 9, 10].includes(k)) {
      assert.equal(s.effect.kind, k);
      assert.equal(s.effect.time, 6);
      assert.equal(s.effect.team, 0);
    }
    if (k === 7) assert.equal(s.ball.owner, 7);
    if (k === 8) assert.equal(s.players[s.ball.owner].team, 0);
    if (k === 11) assert.equal(s.players[7].health, 100);
    if (k === 12) {
      assert.ok(s.players[16].stun > 0);
      assert.equal(s.ball.owner, -1);
    }
  }
});
test("freeze, reversed input, slow, boost, weaken and manic affect actual movement", () => {
  for (const [k, owner, sign, ratio] of [
    [1, 16, 0, 0],
    [2, 16, -1, 1],
    [3, 16, 1, 1],
    [4, 7, 1, 1.4],
    [5, 16, 1, 1.4],
    [6, 16, 1, 1],
  ]) {
    const s = isolated();
    pickup(s, owner, k);
    const before = s.players[7].x;
    step(s, dt, { x: 1 });
    assert.ok(
      Math.abs(s.players[7].x - before - sign * 5 * velocityUnit * dt * ratio) <
        1e-8,
      `power ${k}`,
    );
  }
});
test("temporary power expires, replacement removes old effect, goal reset keeps damage", () => {
  const s = initial();
  pickup(s, 7, 1);
  pickup(s, 16, 10);
  assert.ok(!active(s, 1));
  assert.ok(active(s, 10, 1));
  matchClock(s, 6);
  assert.equal(s.effect.kind, 0);
  s.players[7].health = 45;
  s.tick = 1;
  resetPitch(s);
  assert.equal(s.players[7].health, 45);
});
test("shield prevents tackle and zap damage; heal restores current player", () => {
  const s = initial();
  pickup(s, 16, 10);
  assert.equal(damage(s, 7, 16), false);
  pickup(s, 7, 12);
  assert.equal(s.players[16].health, 100);
  s.effect.time = 0;
  damage(s, 7, 16);
  assert.ok(s.players[16].health < 100);
  pickup(s, 16, 11);
  assert.equal(s.players[16].health, 100);
});
test("goal door blocks the correct goal before and after halftime", () => {
  for (const period of [1, 2]) {
    const s = isolated();
    s.period = period;
    pickup(s, 7, 9);
    const x = period === 1 ? -20.9 : 20.9;
    Object.assign(s.ball, { x, z: 0, h: 1, vx: Math.sign(x) * 24, owner: -1 });
    // Contact is evaluated on the tick after crossing the boundary.
    step(s, dt, {});
    step(s, dt, {});
    assert.equal(s.score[1], 0);
    assert.ok(s.ball.vx * Math.sign(x) < 0);
    assert.ok(!goalBlocked(s, -x, period === 1 ? 1 : -1));
  }
});
test("four warp entrances preserve height and travel direction; high ball rebounds normally", () => {
  for (const x of [-8, 8])
    for (const sign of [-1, 1]) {
      const s = isolated();
      Object.assign(s.ball, {
        x,
        z: 11.1 * sign,
        h: 1,
        vz: 24 * sign,
        vx: 0,
        owner: -1,
        lastTouch: 7,
      });
      // Contact is evaluated on the tick after crossing the boundary.
      step(s, dt, {});
      step(s, dt, {});
      assert.ok(s.ball.z * sign < -10);
      assert.equal(Math.sign(s.ball.vz), sign);
      assert.equal(s.event.kind, 12);
    }
  const s = isolated();
  Object.assign(s.ball, { x: 8, z: 11.1, h: 3, vz: 24, owner: -1 });
  // Contact is evaluated on the tick after crossing the boundary.
  step(s, dt, {});
  step(s, dt, {});
  assert.ok(s.ball.vz < 0);
  assert.ok(s.ball.z > 0);
});
test("electro-bounce gains multiplier charges and knocks an enemy down instead of catching", () => {
  const s = isolated();
  s.multiplier = 2;
  Object.assign(s.ball, {
    x: (304 * 22.4) / 576,
    z: 11.3,
    h: 1,
    lastTouch: 7,
    electricBudget: 3,
  });
  assert.equal(sideFeature(s), true);
  assert.equal(s.ball.electric, 3);
  Object.assign(s.players[16], { x: 5, z: 2, stun: 0, health: 100 });
  Object.assign(s.ball, {
    x: 5,
    z: 2,
    vx: 0,
    vz: 0,
    vh: 0,
    owner: -1,
    lock: 0,
  });
  step(s, dt, {}, [true, true]);
  assert.equal(s.ball.owner, -1);
  assert.equal(s.ball.electric, 2);
  assert.equal(s.ball.electricBudget, 2);
  assert.ok(s.players[16].stun > 0);
  s.players[16].stun = 0;
  pickup(s, 16, 10);
  Object.assign(s.ball, { x: s.players[16].x, z: s.players[16].z });
  step(s, dt, {}, [true, true]);
  assert.equal(s.ball.owner, 16);
  // The original returns before clearing charge when the caught ball is stationary.
  assert.equal(s.ball.electric, 2);
  assert.equal(s.ball.charged, true);
});
test("fatal falls finish before medical scoring, clock pause and reserve rotation", () => {
  const s = initial();
  for (const p of s.players) p.stun = 100;
  for (const item of s.pickups) item.wait = 100;
  for (const points of [10, 20]) {
    Object.assign(s.players[16], { health: 1, stun: 0 });
    damage(s, 7, 16);
    assert.equal(s.score[0], points - 10);
    assert.equal(s.players[16].injury, 0);
    for (let frame = 0; frame < 34; frame++)
      step(s, 1 / 25, {}, [false, false]);
    assert.equal(s.players[16].injury, 0);
    step(s, 1 / 25, {}, [false, false]);
    assert.equal(s.players[16].injury, 1);
    assert.equal(s.score[0], points);
    const time = s.time;
    for (let frame = 0; s.medical && frame < 1000; frame++)
      step(s, 1 / 25, {}, [false, false]);
    assert.equal(s.time, time);
    assert.equal(s.players[16].health, 100);
    assert.equal(s.reserves[1], 3);
  }
});
test("pickups are collected once, respawn, randomize powers and expose equipment", () => {
  const s = isolated(),
    p = s.pickups[0];
  Object.assign(p, { x: 4, z: 0, wait: 0, kind: 13 });
  featureStep(s, dt);
  assert.equal(s.credits[0], 10);
  featureStep(s, dt);
  assert.equal(s.credits[0], 10);
  assert.ok(p.wait > 0);
  const kinds = new Set();
  for (let n = 0; n < 256; n++) {
    spawnPickup(s, 0);
    kinds.add(s.pickups[0].kind);
  }
  assert.equal(kinds.size, 12);
  for (let k = 14; k <= 21; k++) {
    pickup(s, 7, k);
    assert.equal(s.players[7].gear, k);
  }
});
test("speed/throw equipment changes performance and equipment is lost on a hit", () => {
  const s = initial();
  pickup(s, 7, 17);
  assert.equal(movementSpeed(s.players[7], false), 7 * velocityUnit);
  pickup(s, 7, 18);
  throwBall(s, 7, false);
  assert.equal(s.ball.vx, 8 * velocityUnit);
  assert.equal(s.ball.speedTimer, 125);
  damage(s, 16, 7);
  assert.equal(s.players[7].gear, 0);
});

test("transport uses roster slot eight in either half, with no fallback for a fallen target", () => {
  for (const team of [0, 1])
    for (const period of [1, 2]) {
      const s = initial();
      s.period = period;
      const target = team * 9 + 8;
      s.players[target].x = 0;
      s.players[team * 9 + 7].x = 20;
      pickup(s, team * 9 + 7, 8);
      assert.equal(s.ball.owner, target);
      s.ball.owner = 1;
      s.players[target].stun = 1;
      pickup(s, team * 9 + 7, 8);
      assert.equal(s.ball.owner, 1);
    }
});

test("floor items require the selected grounded player and use team-order priority", () => {
  for (const kind of [1, 13, 17]) {
    const s = initial();
    s.players.forEach((p) => {
      p.x = 10;
      p.z = 10;
    });
    const item = s.pickups[0];
    Object.assign(item, { kind, x: 0, z: 0, wait: 0, life: 14 });
    Object.assign(s.players[6], { x: 0, z: 0 });
    featureStep(s, dt);
    assert.equal(item.wait, 0, "unselected player cannot collect");
    Object.assign(s.players[7], { x: 0, z: 0, action: 2 });
    featureStep(s, dt);
    assert.equal(item.wait, 0, "jump cannot collect");
    s.players[7].action = 0;
    Object.assign(s.players[16], { x: 0, z: 0 });
    featureStep(s, dt);
    assert.ok(item.wait > 0);
    if (kind === 1) assert.equal(s.effect.team, 0);
    if (kind === 13) assert.deepEqual(s.credits, [10, 0]);
    if (kind === 17) assert.equal(s.players[7].gear, 17);
  }
});
